src/api/js/src/high-level/high-level.ts

```diff
@@ -0,0 +1,1636 @@
+
+const FALLBACK_PRECISION = 17;
+
+function isCoercibleRational(obj: any): obj is CoercibleRational {
+  // prettier-ignore
+  const r = (
+    (obj !== null &&
+      (typeof obj === 'object' || typeof obj === 'function')) &&
+    (obj.numerator !== null &&
+      (typeof obj.numerator === 'number' || typeof obj.numerator === 'bigint')) &&
+    (obj.denominator !== null &&
+      (typeof obj.denominator === 'number' || typeof obj.denominator === 'bigint'))
+  );
+  r &&
+    assert(
+      (typeof obj.numerator !== 'number' || Number.isSafeInteger(obj.numerator)) &&
+        (typeof obj.denominator !== 'number' || Number.isSafeInteger(obj.denominator)),
+      'Fraction numerator and denominator must be integers',
+    );
+  return r;
+}
+
+export function createApi(Z3: Z3Core): Z3HighLevel {
+  // TODO(ritave): Create a custom linting rule that checks if the provided callbacks to cleanup
+  //               Don't capture `this`
+  const cleanup = new FinalizationRegistry<() => void>(callback => callback());
+
+  function enableTrace(tag: string) {
+    Z3.enable_trace(tag);
+  }
+
+  function disableTrace(tag: string) {
+    Z3.disable_trace(tag);
+  }
+
+  function getVersion() {
+    return Z3.get_version();
+  }
+
+  function getVersionString() {
+    const { major, minor, build_number } = Z3.get_version();
+    return `${major}.${minor}.${build_number}`;
+  }
+
+  function getFullVersion() {
+    return Z3.get_full_version();
+  }
+
+  function openLog(filename: string) {
+    return Z3.open_log(filename);
+  }
+
+  function appendLog(s: string) {
+    Z3.append_log(s);
+  }
+
+  function setParam(key: string, value: any): void;
+  function setParam(params: Record<string, any>): void;
+  function setParam(key: string | Record<string, any>, value?: any) {
+    if (typeof key === 'string') {
+      Z3.global_param_set(key, value.toString());
+    } else {
+      assert(value === undefined, "Can't provide a Record and second parameter to set_param at the same time");
+      Object.entries(key).forEach(([key, value]) => setParam(key, value));
+    }
+  }
+
+  function resetParams() {
+    Z3.global_param_reset_all();
+  }
+
+  function getParam(name: string) {
+    return Z3.global_param_get(name);
+  }
+
+  function isContext(obj: unknown): obj is Context {
+    return obj instanceof ContextImpl;
+  }
+
+  class ContextImpl {
+    declare readonly __typename: Context['__typename'];
+
+    readonly ptr: Z3_context;
+    readonly name: string;
+
+    constructor(name: string, params: Record<string, any> = {}) {
+      const cfg = Z3.mk_config();
+      Object.entries(params).forEach(([key, value]) => Z3.set_param_value(cfg, key, value.toString()));
+      const context = Z3.mk_context_rc(cfg);
+
+      this.ptr = context;
+      this.name = name;
+
+      Z3.set_ast_print_mode(this.ptr, Z3_ast_print_mode.Z3_PRINT_SMTLIB2_COMPLIANT);
+      Z3.del_config(cfg);
+
+      this.AstRef = AstRefImpl.bind(AstRefImpl, this) as any;
```

Comment:
You have "Solver" and "AstRef".
Why not just "Ast"?
The Python API defines Ast as a primitive Python object. The AstRef class wraps around a pointer object. It contains an Ast field.
I don't see where the js bindings refer to an Ast, so this could be used instead of AstRef.